py/dml/ctree.py

```diff
@@ -4297,6 +4307,191 @@ def read(self):
 
 mkHookSendApply = HookSendApply
 
+class TemplatesRef(NonValue):
+    '''Reference to the 'templates' param of an object, used for
+       template-qualified method implementation calls'''
+    @auto_init
+    def __init__(self, site, node, indices): pass
+
+    def __str__(self):
+        return f'{self.node.logname(self.indices)}.templates'
+
+mkTemplatesRef = TemplatesRef
+
+class TemplatesSubRef(NonValue):
+    '''A subreference of the 'templates' param of an object, specifying the
+       template of the template-qualified method implementation call'''
+    @auto_init
+    def __init__(self, site, templates_ref, template_name): pass
+
+    def __str__(self):
+        return f'{self.templates_ref}.{self.template_name}'
+
+def mkTemplatesSubRef(site, templates_ref, template_name):
+    tmpl = dml.globals.templates.get(template_name)
+    if tmpl is None:
+        raise ENTMPL(site, template_name)
+    if tmpl not in templates_ref.node.templates:
+        raise ETQMIC(site, templates_ref.node.ident, template_name)
+    return TemplatesSubRef(site, templates_ref, template_name)
+
+def mkTemplateQualifiedMethodRef(site, templates_subref, method_name):
+    node = templates_subref.templates_ref.node
+    indices = templates_subref.templates_ref.indices
+
+    template_name = templates_subref.template_name
+
+    rank_to_candidate = {}
+
+    # Add any shared method implementation as a possible candidate
+    trait = dml.globals.traits.get(template_name)
+    shared_method = None
+    if trait is not None:
+        for impl_trait in trait.method_impl_traits.get(method_name, []):
+            impl_template = dml.globals.templates[impl_trait.name]
+            shared_method = impl_trait.method_impls[method_name]
+            rank_to_candidate[impl_template.spec.rank] = (
+                impl_template, shared_method)
+
+
+    studied_templates = set()
+    pending_templates = Set((dml.globals.templates[template_name],)
+                            # don't bother going through regular methods
+                            # if if some shared method impl. is
+                            # unoverridable (can only be one if so)
+                            * (shared_method is None
+                               or shared_method.overridable))
```

Comment:
I think explicit control flow would improve readability drastically:
```
if not shared or overridable:
   pending = [one_element]
   studied = set()
   while pending:
      ...      
```

Also, possibly more straightforward to keep pending_templates a list, and eliminate duplicates on pop instead (I prefer native Python types when I can). But I don't insist.

Reply:
God no, that incurs another indendation level

Reply:
> Also, possibly more straightforward to keep pending_templates a list, and eliminate duplicates on pop instead (I prefer native Python types when I can). But I don't insist.
I could be OK with that.

Reply:
> God no, that incurs another indendation level

I object! high-level control flow is more important than indent level. If indent level is too deep, then it's time to factor out a function. "repeat until this list, that starts at size 1 or 0 depending on XYZ, is empty" is not a natural way of thinking.

Reply:
uurggh. Fine. I wanted to avoid indentation due to the long name of `get_potential_method_impl_details`, which would incur horrible-to-read-and-resolve linebreaks, but it turns out I can just _barely_ avoid the worst cases.